### PR TITLE
Add module expressions

### PR DESCRIPTION
**Pull request by Peter554**
Context: https://github.com/seddonym/import-linter/issues/252

This PR adds module expressions to grimp. By moving module expressions into grimp, in particular into the rust code, we will be able to speed up lots of the import linter code that works with module expressions (import linter now supports module expressions in quite a few places).

### CODE REVIEW

File: docs/usage.rst [hunk 1]
```diff
@@ -517,5 +517,25 @@ Methods for manipulating the graph
     :param str module: The name of a module, for example ``'mypackage.foo'``.
     :return: bool
 
+.. _module_expressions:
+
+Module expressions
```
Comment (Peter554) [inline docs/usage.rst:553]: 🍷 This text pretty much copied from import linter docs here https://import-linter.readthedocs.io/en/stable/contract_types.html#wildcards

File: rust/src/graph/hierarchy_queries.rs [hunk 2]
```diff
@@ -40,11 +42,52 @@ impl Graph {
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::graph::ModuleIterator;
+    use std::collections::HashSet;
+
+    #[test]
+    fn test_find_matching_modules() {
```
Comment (Peter554) [inline rust/src/graph/hierarchy_queries.rs:71]: 🍷  The test here is quite light, just a sanity check - we don't need to check again here how all the different types of module expressions work since that is already tested thoroughly in the dedicated unit tests (see previous commit).

File: tests/unit/adaptors/graph/test_hierarchy.py [hunk 3]
```diff
@@ -115,3 +115,22 @@ def test_find_descendants_works_if_modules_added_in_different_order():
...
+class TestFindMatchingModules:
+    def test_finds_matching_modules(self):
+        graph = ImportGraph()
+        graph.add_module("foo")
+        graph.add_module("foo.bar")
+        graph.add_module("foo.bar.baz")
+
+        assert graph.find_matching_modules("foo") == {"foo"}
+        assert graph.find_matching_modules("foo.*") == {"foo.bar"}
+        assert graph.find_matching_modules("foo.**") == {"foo.bar", "foo.bar.baz"}
```
Comment (Peter554) [inline tests/unit/adaptors/graph/test_hierarchy.py:129]: 🍷 This test basically duplicates the rust test. For me it's okay to keep both.

File: src/grimp/application/ports/graph.py [hunk 4]
```diff
@@ -32,6 +32,16 @@ def modules(self) -> Set[str]:
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def find_matching_modules(self, expression: str) -> Set[str]:
+        """
+        Find all modules matching the passed expression.
+
+        See https://grimp.readthedocs.io/en/stable/usage.html#module-expressions
+        for a description of module expression syntax.
```
Comment (Peter554) [inline src/grimp/application/ports/graph.py:46]: 🍷  Rather than duplicate all the information here about "what are module expressions" (and also in `find_matching_direct_imports`), I thought I'd just refer to the docs. Is that sensible?

File: rust/src/graph/direct_import_queries.rs [hunk 5]
```diff
@@ -54,4 +55,58 @@ impl Graph {
             None => &EMPTY_IMPORT_DETAILS,
         }
     }
+
+    pub fn find_matching_direct_imports(
+        &self,
+        importer_expression: &ModuleExpression,
+        imported_expression: &ModuleExpression,
+    ) -> FxHashSet<(ModuleToken, ModuleToken)> {
```
Comment (Peter554) [inline rust/src/graph/direct_import_queries.rs:63]: 🍷 `(ModuleToken, ModuleToken)` could probably be made a bit nicer with a struct, but it's consistent with what we're doing in the rest of the rust code for now, so I think I'd leave it like this here and leave the introduction of that struct to a separate change.

File: rust/src/graph/direct_import_queries.rs [hunk 6]
```diff
@@ -54,4 +55,58 @@ impl Graph {
...
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_find_matching_direct_imports() {
```
Comment (Peter554) [inline rust/src/graph/direct_import_queries.rs:89]: 🍷 ♻️ https://github.com/seddonym/grimp/pull/188#discussion_r1966455262

File: tests/unit/adaptors/graph/test_direct_imports.py [hunk 7]
```diff
@@ -265,3 +266,45 @@ def test_returns_only_relevant_imports(self):
...
+            line_number=1,
+            line_contents="...",
+        )
+
+        assert graph.find_matching_direct_imports(
+            importer_expression="pkg.animals.*", imported_expression="pkg.colors.*"
+        ) == [
+            {"importer": "pkg.animals.cat", "imported": "pkg.colors.ginger"},
+            {"importer": "pkg.animals.dog", "imported": "pkg.colors.golden"},
+        ]
```
Comment (Peter554) [inline tests/unit/adaptors/graph/test_direct_imports.py:292]: 🍷 ♻️ https://github.com/seddonym/grimp/pull/188#discussion_r1966455346

File: src/grimp/application/ports/graph.py [hunk 8]
```diff
@@ -187,6 +192,18 @@ def get_import_details(self, *, importer: str, imported: str) -> List[DetailedIm
...
 
+    @abc.abstractmethod
+    def find_matching_direct_imports(
+        self, *, importer_expression: str, imported_expression: str
+    ) -> List[Import]:
+        """
+        Find all direct imports matching the passed expressions.
+
+        See https://grimp.readthedocs.io/en/stable/usage.html#module-expressions
+        for a description of module expression syntax.
```
Comment (Peter554) [inline src/grimp/application/ports/graph.py:203]: 🍷 ♻️ https://github.com/seddonym/grimp/pull/188#discussion_r1966455669

File: src/grimp/application/ports/graph.py [hunk 9]
```diff
@@ -9,6 +9,11 @@
 from grimp.domain.valueobjects import Layer
 
 
+class Import(TypedDict):
```
Comment (Peter554) [inline src/grimp/application/ports/graph.py:12]: I'm wondering about the name here - the name is to be contrasted with the `DetailedImport` typed dict a few lines below. If "with line number and contents" is a `DetailedImport` then "without line number and contents" is just an `Import`? Other ideas might be `CompactImport` or `ImportSummary`.

File: rust/src/graph/pathfinding.rs [hunk 10]
```diff
@@ -28,7 +28,8 @@ pub fn find_reach(
     &seen.into_iter().collect::<FxHashSet<_>>() - from_modules
 }
 
-pub fn find_shortest_path_bidirectional(
+/// Finds the shortest path, via a bidirectional BFS.
+pub fn find_shortest_path(
```
Comment (Peter554) [inline rust/src/graph/pathfinding.rs:32]: 🍷 Sneaking this little change into the PR - it's not related to the rest of the changes, I can remove it if you like.

File: tests/unit/adaptors/graph/test_hierarchy.py (hunk 3 above)
Comment (seddonym) [reply tests/unit/adaptors/graph/test_hierarchy.py:129]: I think a good rule of thumb is that the Python tests should be comprehensive, and if we want to add some Rust tests to help too (even if they duplicate) then that's fine. Do you agree?

**Review by seddonym — COMMENTED** [review]
Just leaving some early feedback - I haven't yet gone over everything in detail.

**Review by seddonym — COMMENTED** [review]
Another round of comments.

The commits are on the large side for reviewing, by the way...

File: src/grimp/application/ports/graph.py (hunk 9 above)
Comment (Peter554) [reply src/grimp/application/ports/graph.py:12]: > Should DetailedImport inherit Import?

Done

File: rust/src/module_expressions.rs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+use std::fmt::Display;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref MODULE_EXPRESSION_PATTERN: Regex =
+        Regex::new(r"^(\w+|\*{1,2})(\.(\w+|\*{1,2}))*$").unwrap();
+}
+
+#[derive(Debug, Clone)]
+pub struct ModuleExpression {
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs:17]: Added a doc comment here now.

File: rust/src/module_expressions.rs [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        } else {
+            self.expression == module_name
+        }
+    }
+
+    fn create_pattern(expression: &str) -> GrimpResult<Regex> {
+        let mut pattern_parts = vec![];
+
+        for part in expression.split(".") {
+            if part == "*" {
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs:60]: I tried writing a comment, but found that the resulting comment was more confusing than just reading the code/regex. Especially the `"**"` branch is kind of hard to describe, the regex says it best I think.

File: docs/usage.rst [hunk 13]
```diff
@@ -121,6 +121,16 @@ Methods for analysing the module tree
     :raises: ``ValueError`` if the module is a squashed module, as by definition it represents both itself and all
       of its descendants.
 
+.. py:function:: ImportGraph.find_matching_modules(expression)
```
[the remark this answers is not in the document]
Comment (Peter554) [reply docs/usage.rst:124]: From my side I don't see a need, and it does make the API more complex. Unless you are wanting to query a really massive set of expressions, I think there won't be much performance benefit over just using a `for` loop in the calling code. So I think I'd prefer to leave the API like this. But of course it's up to you, just let me know.

File: rust/src/graph/hierarchy_queries.rs [hunk 14]
```diff
@@ -40,11 +42,52 @@ impl Graph {
...
         }
         descendants.into_iter()
     }
+
+    pub fn find_matching_modules(&self, expression: &ModuleExpression) -> impl ModuleIterator {
+        let interner = MODULE_NAMES.read().unwrap();
+        let modules: FxHashSet<_> = self
+            .modules
+            .values()
+            .filter(|m| expression.is_match(interner.resolve(m.interned_name).unwrap()))
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/hierarchy_queries.rs:58]: I think probably, but I can check it with the benchmark 😁

File: src/grimp/application/ports/graph.py [hunk 15]
```diff
@@ -32,6 +37,16 @@ def modules(self) -> Set[str]:
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def find_matching_modules(self, expression: str) -> Set[str]:
+        """
+        Find all modules matching the passed expression.
```
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:41]: Yes, done it

File: rust/src/graph/hierarchy_queries.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/hierarchy_queries.rs:58]: This seems to make it slower 👇 The cost of multiple threads outweighs the benefits.

```
------ benchmark 'test_find_matching_modules': 2 tests -------
Name (time in ms)                               Mean
--------------------------------------------------------------
test_find_matching_modules (0173_48ec860)     1.3820 (1.0)
test_find_matching_modules (NOW)              5.1019 (3.69)
--------------------------------------------------------------
```

File: rust/src/graph/direct_import_queries.rs [hunk 16]
```diff
@@ -54,4 +55,58 @@ impl Graph {
...
+        self.imports
+            .iter()
+            .flat_map(|(importer, imports)| {
+                imports
+                    .iter()
+                    .cloned()
+                    .map(move |imported| (importer, imported))
+            })
+            .filter(|(importer, imported)| {
+                let importer = self.get_module(*importer).unwrap();
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/direct_import_queries.rs:74]: We don't pass modules, we pass module expressions. So if we pass an module expressions that is just an exact module name that does not exist, then this module expression will simply not match on anything. So we'll get an empty list as a result (no error will be raised). That seems sensible to me. What do you think?

File: rust/src/graph/direct_import_queries.rs [hunk 17]
```diff
@@ -54,4 +55,58 @@ impl Graph {
...
+                let importer = interner.resolve(importer.interned_name).unwrap();
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/direct_import_queries.rs:75]: I think this is actually quite idiomatic in rust. And we're doing similar in other parts of the codebase. So I'm inclined to leave it.

* https://www.ntietz.com/blog/rust-shadowing-idiomatic/
* https://users.rust-lang.org/t/inquiring-about-idiomatic-usage-of-shadowing/6177/2

File: src/grimp/application/ports/graph.py [hunk 18]
```diff
@@ -177,6 +192,18 @@ def get_import_details(self, *, importer: str, imported: str) -> List[DetailedIm
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def find_matching_direct_imports(
+        self, *, importer_expression: str, imported_expression: str
+    ) -> List[Import]:
```
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:220]: Yes added a comment in the docstring that this is an ordered List

```
Returns:
    An ordered list of direct imports matching the expressions (ordered alphabetically).
```

That should make it obvious why a set can't work here (sets aren't ordered).

[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:220]: Yes, added this in the docstring.

File: src/grimp/application/ports/graph.py [hunk 19]
```diff
@@ -177,6 +192,18 @@ def get_import_details(self, *, importer: str, imported: str) -> List[DetailedIm
...
+        """
+        Find all direct imports matching the passed expressions.
```
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:222]: Done

File: src/grimp/application/ports/graph.py (hunk 4 above)
Comment (Peter554) [reply src/grimp/application/ports/graph.py]: I've duplicated the description of module names in the code now in the docstring of `find_matching_modules` (so the description is once in the code, once in the docs).

File: src/grimp/application/ports/graph.py [hunk 20]
```diff
@@ -177,6 +192,18 @@ def get_import_details(self, *, importer: str, imported: str) -> List[DetailedIm
         """
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def find_matching_direct_imports(
+        self, *, importer_expression: str, imported_expression: str
```
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:219]: Similar to https://github.com/seddonym/grimp/pull/188#discussion_r1968349458, from my side I think probably not. I think if you want to check multiple import expressions then a `for` loop by the caller should be sufficient. Unless you have a really massive list of import expressions to query, there won't be much performance benefit to doing this in a single call.

You also mention the term "import expressions" here - as the changes are currently in this PR, grimp doesn't have a concept of import expressions. I.e. there is no `grimp.ImportExpression` in the public API. I didn't add this, because copying from our slack conversation:
> should we take the module expressions as str or should there be a grimp.ModuleExpression (and grimp.ImportExpression?)
> ...
> I think input types should be strings. That fits API style rest of grimp.

[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:219]: Again, it's up to you though of course, just let me know.

File: tests/unit/adaptors/graph/test_hierarchy.py (hunk 3 above)
Comment (Peter554) [reply tests/unit/adaptors/graph/test_hierarchy.py:129]: > Do you agree?

I think I could be persuaded either way. In the sense of a testing pyramid, one might argue that the rust tests should be the most comprehensive, since this is the lowest layer of the code. All the python tests integrate the rust code, so by definition are higher up the pyramid, so we'd expect less comprehensive tests there. 

That said, this isn't a normal testing pyramid, and the preexisting code puts most of the tests in python. 

In summary I think it's good to have both, certainly with comprehensive python tests. I'll flesh out the python tests a bit more.

File: rust/src/module_expressions.rs [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+    type Err = GrimpError;
+
+    fn from_str(expression: &str) -> Result<Self, Self::Err> {
+        if !MODULE_EXPRESSION_PATTERN.is_match(expression) {
+            return Err(GrimpError::InvalidModuleExpression(expression.to_owned()));
+        }
+
+        let contains_wildcard = expression.contains('*');
+
+        let pattern = if contains_wildcard {
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs]: Agree, I removed the `Option` now

File: tests/unit/adaptors/graph/test_hierarchy.py [hunk 22]
```diff
@@ -115,3 +115,30 @@ def test_find_descendants_works_if_modules_added_in_different_order():
...
+
+
+class TestFindMatchingModules:
+    def test_finds_matching_modules(self):
+        graph = ImportGraph()
+        graph.add_module("foo")
+        graph.add_module("foo.bar")
+        graph.add_module("foo.bar.baz")
+
+        assert graph.find_matching_modules("foo") == {"foo"}
```
[the remark this answers is not in the document]
Comment (Peter554) [reply tests/unit/adaptors/graph/test_hierarchy.py]: I'm reasonably confident that we have similar coverage - the rust test cases within `module_expressions.rs` cover all of that I think. There aren't many tests at the level of `find_matching_modules` and `find_matching_direct_imports`. Most of the testing is done as unit tests on the `ModuleExpression` i.e. `test_parse` and `test_is_match`.

File: rust/src/graph/direct_import_queries.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/direct_import_queries.rs]: On second thought I decided I would change this. Done.

File: src/grimp/application/ports/graph.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:219]: Ah I think I might have misunderstood you - you are suggesting to pass import expressions in as strings of the form `"a.b -> c.d"`? I think I'm slightly against that. This kind of expression feels a bit too close to the interface layer to me. E.g. import linter might care about import expressions of that form, and have logic for splitting on the `"->"` and stripping whitespace, but I think it's not a concept grimp needs to know about.

File: docs/usage.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply docs/usage.rst:124]: I think it's also a bit inconsistent with the rest of the grimp API - we're not generally doing bulk methods like this I think.

File: tests/unit/adaptors/graph/test_hierarchy.py (hunk 3 above)
Comment (seddonym) [reply tests/unit/adaptors/graph/test_hierarchy.py]: Thinking back to when we swapped in your Rust implementation, it was a really useful property of the test suite that  there was thorough test coverage of the Python API - we just deleted all the Rust tests and still felt confident. If coverage had been split across the two languages then we wouldn't have been able to do that.

So I think focusing on the Python tests is worthwhile - we can treat the Rust tests more lightly as tests to help us as we go along, but more disposable.

Arguably the Python unit tests are still at the base of the testing pyramid - they're fast and give us lots of signal about granular behaviour.

File: tests/unit/adaptors/graph/test_hierarchy.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (seddonym) [reply tests/unit/adaptors/graph/test_hierarchy.py]: As per [the other discussion](https://github.com/seddonym/grimp/pull/188#discussion_r1966455346), would you mind making sure the behaviour is covered by the Python tests, too? Or I can do it once everything else is ready on the PR.

File: rust/src/graph/direct_import_queries.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (seddonym) [reply rust/src/graph/direct_import_queries.rs]: I prefer it personally, in this case at least - makes it clearer what the code is doing. Thanks.

File: rust/src/module_expressions.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (seddonym) [reply rust/src/module_expressions.rs]: Great, looks good! Thanks.

**Review by seddonym — CHANGES_REQUESTED** [review]
Ok I've finished the full review. It's looking great in general, thanks for your efforts!

I don't think there's much more to do on this, mainly it's just a few tweaks to comments.

Before merging to master, there are two blockers:

- Making sure the Python test coverage is comprehensive.
- Figuring out whether to represent import expressions as two-tuples or arrow-strings (let's discuss).

I'd be happy to merge this into a branch in my repo if it would help - then we could both work on the remainder. Let me know what you prefer.

File: src/grimp/application/ports/graph.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:219]: >  Do any of those arguments change your position?

Somewhat convincing, but I think I'm still -0 on the arrow strings.

Suppose I already have two module expressions as strings. Would you expect callers to do this (it looks a bit ugly IMO)?

```py
# Suppose I start with
importer_expression = ...
imported_expression = ...

# Then I have to do this..?
find_matching_direct_imports(f"{importer_expression} -> {imported_expression}")
``` 

We know that import linter will be working from these arrow formatted strings from the start, but for people using grimp without import linter it seems a bit odd to me to make them format their arguments in this way.

File: rust/src/graph/direct_import_queries.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/graph/direct_import_queries.rs:74]: Yes I agree.

File: tests/unit/adaptors/graph/test_hierarchy.py (hunk 3 above)
Comment (Peter554) [reply tests/unit/adaptors/graph/test_hierarchy.py]: Yes okay I think you're right 👍

File: src/grimp/application/ports/graph.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply src/grimp/application/ports/graph.py:220]: True, this could probably be made a little clearer, let's change this.

File: rust/src/module_expressions.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs:68]: > do you think you could explain your thinking behind it 

When I read the regex my thinking is :

* `[^\.]+`: One of more non-`.` characters
* `(?:\.[^\.]+)*?` Zero or more groups of (lazy/ungreedy)
  * `?:`: Non-capturing group
  * `\.`: The literal `.`
  * `[^\.]+`:  One of more non-`.` characters

It there any less technical way to write this?

File: rust/src/module_expressions.rs [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+                pattern_parts.push(part.replace("*", r"[^\.]+"))
+            } else if part == "**" {
+                pattern_parts.push(part.replace("**", r"[^\.]+(?:\.[^\.]+)*?"))
+            } else {
+                pattern_parts.push(part.to_owned());
+            }
+        }
+
+        Regex::new(&(r"^".to_owned() + &pattern_parts.join(r".") + r"$"))
+            .map_err(|_| GrimpError::InvalidModuleExpression(expression.to_owned()))
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs:78]: Yes true - I'd be fine to panic here instead.

File: rust/src/module_expressions.rs [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+impl AsRef<str> for ModuleExpression {
+    fn as_ref(&self) -> &str {
+        &self.expression
+    }
+}
+
+impl FromStr for ModuleExpression {
+    type Err = GrimpError;
+
+    fn from_str(expression: &str) -> Result<Self, Self::Err> {
```
[the remark this answers is not in the document]
Comment (Peter554) [reply rust/src/module_expressions.rs:38]: I _think_ `FromStr` is simply more idiomatic than `new` for this case, since we are parsing our struct from a string. `new` could also work though 🤷

File: rust/src/module_expressions.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (seddonym) [reply rust/src/module_expressions.rs:68]: One level up - why are we doing it like this? Or maybe a couple of examples?

Perhaps a way in is to think about how this behaves with multiple sets of `**` - e.g. what's the behaviour of `mypackage.**.blue.green.**`?

File: rust/src/module_expressions.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (seddonym) [reply rust/src/module_expressions.rs:38]: Cool, just curious.

**Review by seddonym — APPROVED** [review]
Merging to the `add-module-expressions` working branch in the origin repo.
